_posts/2024-09-05-digging-into-struct-initialization-performance.md

```diff
@@ -0,0 +1,254 @@
+
+> [!TIP]
+> Take away: the design of `Digest` isn’t playing nice with algorithms that need lots of memory, at least for small strings.
+
+We only need the hexstring to be allocated in the HEAP, the 2KB are allocated and thrown away so we can try to put them on the stack and call the C functions directly. Let’s verify if it improves the situation.
+
+```crystal
+require "benchmark"
+require "blake3"
+require "digest/sha256"
+
+def blake3_hexstring(data)
+  hasher = uninitialized UInt8[1912]
+  hashsum = uninitialized UInt8[32]
+  Digest::Blake3::Lib.init pointerof(hasher)
+  Digest::Blake3::Lib.update pointerof(hasher), data, data.bytesize
+  Digest::Blake3::Lib.final pointerof(hasher), hashsum.to_slice, hashsum.size
+  hashsum.to_slice.hexstring
+end
+
+Benchmark.ips do |x|
+  bytes = Random::Secure.random_bytes(32)
+  x.report("SHA256") { Digest::SHA256.hexdigest(bytes) }
+  x.report("Blake3") { blake3_hexstring(bytes) }
+end
+```
+
+```
+SHA256   1.33M (754.01ns) (± 1.09%)   225B/op   3.40× slower
+Blake3   4.51M (221.76ns) (± 2.57%)  80.0B/op        fastest
+```
+
+We now only allocate 80 bytes for each digest (for the hexstring) and BLAKE3 is much faster! We’re far from the 14× claim, but the data to hash is small, and the C library chose the SSE4.1 assembly for my CPU; the AVX512 assembly would likely be faster, but my CPU doesn’t support it.
+
+## Let’s refactor as idiomatic Crystal
+
+With the performance back, I went on with refactoring the shard, wrapping the C functions inside a `struct` so we get a nice, idiomatic and optimized Crystal. At best we can use the struct directly, for example in the one time uses of `Digest::Blake3.hexdigest`. At worst I’ll embed it in the class for the yielding and streaming cases that will need to allocate 2KB in the HEAP, but the longer the message to hash, the less impactful the initial HEAP allocation is.
+
+```crystal
+require "benchmark"
+require "blake3"
+require "digest/sha256"
+
+struct Blake3Hasher
+  def initialize
+    @hasher = uninitialized UInt8[1912]
+    Digest::Blake3::Lib.init(self)
+  end
+
+  def update(data)
+    Digest::Blake3::Lib.update(self, data.to_slice, data.bytesize)
+  end
+
+  def final(hashsum)
+    Digest::Blake3::Lib.final(self, hashsum, hashsum.size)
+  end
+
+  def to_unsafe
+    pointerof(@hasher)
+  end
+end
+
+def blake3_hexstring(data)
+  hasher = Blake3Hasher.new
+  hashsum = uninitialized UInt8[32]
+  hasher.update(data)
+  hasher.final(hashsum.to_slice)
+  hashsum.to_slice.hexstring
+end
+
+Benchmark.ips do |x|
+  bytes = Random::Secure.random_bytes(32)
+  x.report("SHA256") { Digest::SHA256.hexdigest(bytes) }
+  x.report("Blake3") { blake3_hexstring(bytes) }
+end
+```
+
+```
+SHA256   1.34M (744.61ns) (± 0.52%)   225B/op   1.38× slower
+Blake3   1.85M (539.81ns) (± 1.22%)  80.0B/op        fastest
+```
+
+And… the performance gain is almost entirely gone.
```

Comment:
Nitpick: This sounds a bit too much, since it's still much better than the first version in which SHA256 was faster than Blake3.

Reply:
We still go from 4.5M iterations/s to 1.85M, so:

> And... performance is crashing down.